File: behavior/observer.js
/*
* Наблюдатель (Observer, Dependents, Publish-Subscribe) - паттерн поведения объектов.
*
* Определяет зависимость типа "один ко многим" между объектами таким образом,
* что при изменении состояния одного объекта все зависящие от него оповещаются об этом
* и автоматически обновляются.
*
* */


function Subject(){
    /* Субъект */
    this.data = null;
    this.observers = [];
}

Subject.prototype.attach = function(observer){
    /* подписаться на оповещение */

    // проверим класс добавляемого объекта
    if (!observer instanceof Observer){
        throw Error('Неверный класс объекта');
    }

    for(var i = 0; i < this.observers.length; i++){
        if(this.observers[i] === observer){
            return;
        }
    }
    this.observers.push(observer);
};

Subject.prototype.detach = function(observer){
    /* отписаться от оповещения */
    for(var i = 0; i < this.observers.length; i++){
        if(this.observers[i] === observer){
            this.observers.splice(i, 1);
            return;
        }
    }
};

Subject.prototype.getData = function(){
    return this.data;
};

Subject.prototype.setData = function(data){
    this.data = data;
    this.notify(data);
};

Subject.prototype.notify = function(data){
    /* уведомить всех наблюдателей о событии */
    for(var i = 0; i < this.observers.length; i++){
        this.observers[i].update(data);
    }
};

function Observer(name){
    /* Наблюдатель */
    this.name = name;
}

Observer.prototype.update = function(data){
    console.log(this.name, ':', data);
};

var subject = new Subject(),
    observer1 = new Observer('Наблюдатель 1'),
    observer2 = new Observer('Наблюдатель 2');

subject.attach(observer1);
subject.attach(observer2);
subject.setData('данные для наблюдателя');

// Наблюдатель 1 : данные для наблюдателя
// Наблюдатель 2 : данные для наблюдателя

File: behavior/observer.py
# coding: utf-8

"""
Наблюдатель (Observer, Dependents, Publish-Subscribe) - паттерн поведения объектов.

Определяет зависимость типа "один ко многим" между объектами таким образом,
что при изменении состояния одного объекта все зависящие от него оповещаются об этом
и автоматически обновляются.
"""


class Subject(object):
    """Субъект"""
    def __init__(self):
        self._data = None
        self._observers = set()

    def attach(self, observer):
        # подписаться на оповещение
        if not isinstance(observer, ObserverBase):
            raise TypeError()
        self._observers.add(observer)

    def detach(self, observer):
        # отписаться от оповещения
        self._observers.remove(observer)

    def get_data(self):
        return self._data

    def set_data(self, data):
        self._data = data
        self.notify(data)

    def notify(self, data):
        # уведомить всех наблюдателей о событии
        for observer in self._observers:
            observer.update(data)


class ObserverBase(object):
    """Абстрактный наблюдатель"""
    def update(self, data):
        raise NotImplementedError()


class Observer(ObserverBase):
    """Наблюдатель"""
    def __init__(self, name):
        self._name = name

    def update(self, data):
        print '%s: %s' % (self._name, data)


subject = Subject()
subject.attach(Observer('Наблюдатель 1'))
subject.attach(Observer('Наблюдатель 2'))
subject.set_data('данные для наблюдателя')
# Наблюдатель 2: данные для наблюдателя
# Наблюдатель 1: данные для наблюдателя
